src/Init/Data/String/Slice.lean

```diff
@@ -0,0 +1,1352 @@
+  hash := hash
+
+instance : LT Slice where
+  lt x y := x.copy < y.copy
+
+@[extern "lean_slice_dec_lt"]
+instance : DecidableLT Slice :=
+  fun x y => inferInstanceAs (Decidable (x.copy < y.copy))
+
+instance : Ord Slice where
+  compare x y := compareOfLessAndBEq x y
+
+instance : LE Slice where
+  le x y := ¬x < y
+
+instance : DecidableLE Slice :=
+  fun x y => inferInstanceAs (Decidable (¬x < y))
+
+section ForwardPatternUsers
+
+variable {ρ : Type} {σ : Slice → Type}
+variable [∀ s, Std.Iterators.Iterator (σ s) Id (SearchStep s)]
+variable [∀ s, Std.Iterators.Finite (σ s) Id]
+variable [∀ s, Std.Iterators.IteratorLoop (σ s) Id Id]
+
+/--
+Checks whether the slice ({name}`s`) begins with the pattern ({name}`pat`).
+
+This function is generic over all currently supported patterns.
+
+Examples:
+ * {lean}`"red green blue".toSlice.startsWith "red" = true`
+ * {lean}`"red green blue".toSlice.startsWith "green" = false`
+ * {lean}`"red green blue".toSlice.startsWith "" = true`
+ * {lean}`"red green blue".toSlice.startsWith 'r' = true`
+ * {lean}`"red green blue".toSlice.startsWith Char.isLower = true`
+-/
+@[inline]
+def startsWith [ForwardPattern ρ] (s : Slice) (pat : ρ) : Bool :=
+  ForwardPattern.startsWith s pat
+
+inductive SplitIterator (ρ : Type) [ToForwardSearcher ρ σ] where
+  | operating (s : Slice) (currPos : s.Pos) (searcher : Std.Iter (α := σ s) (SearchStep s))
+  | atEnd
+  deriving Inhabited
+
+namespace SplitIterator
+
+variable [ToForwardSearcher ρ σ]
+
+instance [Pure m] : Std.Iterators.Iterator (SplitIterator ρ) m Slice where
+  IsPlausibleStep := fun _ _ => True
+  step := fun ⟨iter⟩ =>
+    match iter with
+    | .operating s currPos searcher =>
+      match Internal.nextMatch searcher with
+      | some (searcher, startPos, endPos) =>
+        let slice := s.replaceStartEnd! currPos startPos
+        let nextIt := ⟨.operating s endPos searcher⟩
+        pure ⟨.yield nextIt slice, by simp⟩
+      | none =>
+        let slice := s.replaceStart currPos
+        pure ⟨.yield ⟨.atEnd⟩ slice, by simp⟩
+    | .atEnd => pure ⟨.done, by simp⟩
+
+-- TODO: Finiteness after we have a notion of lawful searcher
+
+instance [Monad m] [Monad n] : Std.Iterators.IteratorCollect (SplitIterator ρ) m n :=
+  .defaultImplementation
+
+instance [Monad m] [Monad n] : Std.Iterators.IteratorCollectPartial (SplitIterator ρ) m n :=
+  .defaultImplementation
+
+instance [Monad m] [Monad n] : Std.Iterators.IteratorLoop (SplitIterator ρ) m n :=
+  .defaultImplementation
+
+instance [Monad m] [Monad n] : Std.Iterators.IteratorLoopPartial (SplitIterator ρ) m n :=
+  .defaultImplementation
+
+end SplitIterator
+
+/--
+Splits a slice at each subslice that matches the pattern {name}`pat`.
+
+The subslices that matched the pattern are not included in any of the resulting subslices. If
+multiple subslices in a row match the pattern, the resulting list will contain empty strings.
+
+This function is generic over all currently supported patterns.
+
+Examples:
```

Comment:
I feel that this could do with an example illustrating the case of overlapping matches.